Serve a file through a readable stream

I want to send a file without putting the whole file in memory, for that i use futures-fs who creat a readable stream, but i cant give to a Hyper response, with the Body::wrap_stream function:
Code:
```rust
use hyper::service::{ make_service_fn, service_fn };
use hyper::{ Body, Request, Response, Server };
use futures_fs::FsPool;

type GenericError = Box<dyn std::error::Error + Send + Sync>;
type Result<T> = std::result::Result<T, GenericError>;

async fn serve(fs: FsPool, req: Request<Body>) -> Result<Response<Body>> {
	// let mut not_found = Response::default();
	let stream = fs.read("/tmp/streams/hello_world.html", Default::default());
	Ok(Response::new(Body::wrap_stream(stream)))
}

#[tokio::main]
async fn main() -> Result<()> {
    let addr = ([127, 0, 0, 1], 3000).into();
	let fs = FsPool::default();
    let service = make_service_fn(|_| {
		let fs = fs.clone();
		async move {
			Ok::<_, hyper::Error>(service_fn(move |req| {
				serve(fs.clone(), req)
			})) 
		}
		});
    let server = Server::bind(&addr).serve(service);
    server.await?;
    Ok(())
}
```
I got the following error:
```bash
error[E0277]: the trait bound `futures_fs::read::FsReadStream: futures_core::stream::Stream` is not satisfied
   --> src/main.rs:11:37
    |
11  |     Ok(Response::new(Body::wrap_stream(stream)))
    |                                        ^^^^^^ the trait `futures_core::stream::Stream` is not implemented for `futures_fs::read::FsReadStream`
    | 
   ::: /home/tet/.cargo/registry/src/github.com-1ecc6299db9ec823/hyper-0.13.4/src/body/body.rs:159:12
    |
159 |         S: Stream<Item = Result<O, E>> + Send + Sync + 'static,
    |            --------------------------- required by this bound in `hyper::body::body::Body::wrap_stream`

error: aborting due to previous error

```

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -51,6 +51,7 @@ serde_derive = "1.0"
 serde_json = "1.0"
 tokio = { version = "0.2.2", features = ["fs", "macros", "io-std", "rt-util", "sync", "time", "test-util"] }
 tokio-test = "0.2"
+tokio-util = { version = "0.3", features = ["codec"] }
 tower-util = "0.3"
 url = "1.0"
 

--- a/examples/README.md
+++ b/examples/README.md
@@ -39,7 +39,7 @@ pretty_env_logger = "0.4"
 
 * [`params`](params.rs) - A webserver that accept a form, with a name and a number, checks the parameters are presents and validates the input.
 
-* [`send_file`](send_file.rs) - A server that sends back content of files using tokio_fs to read the files asynchronously.
+* [`send_file`](send_file.rs) - A server that sends back content of files using tokio-util to read the files asynchronously.
 
 * [`single_threaded`](single_threaded.rs) - A server only running on 1 thread, so it can make use of `!Send` app state (like an `Rc` counter).
 

--- a/examples/send_file.rs
+++ b/examples/send_file.rs
@@ -1,13 +1,13 @@
 #![deny(warnings)]
 
 use tokio::fs::File;
-use tokio::io::AsyncReadExt;
+
+use tokio_util::codec::{BytesCodec, FramedRead};
 
 use hyper::service::{make_service_fn, service_fn};
 use hyper::{Body, Method, Request, Response, Result, Server, StatusCode};
 
 static INDEX: &str = "examples/send_file_index.html";
-static INTERNAL_SERVER_ERROR: &[u8] = b"Internal Server Error";
 static NOTFOUND: &[u8] = b"Not Found";
 
 #[tokio::main]
@@ -30,9 +30,7 @@ async fn main() {
 
 async fn response_examples(req: Request<Body>) -> Result<Response<Body>> {
     match (req.method(), req.uri().path()) {
-        (&Method::GET, "/") | (&Method::GET, "/index.html") | (&Method::GET, "/big_file.html") => {
-            simple_file_send(INDEX).await
-        }
+        (&Method::GET, "/") | (&Method::GET, "/index.html") => simple_file_send(INDEX).await,
         (&Method::GET, "/no_file.html") => {
             // Test what happens when file cannot be be found
             simple_file_send("this_file_should_not_exist.html").await
@@ -49,26 +47,13 @@ fn not_found() -> Response<Body> {
         .unwrap()
 }
 
-/// HTTP status code 500
-fn internal_server_error() -> Response<Body> {
-    Response::builder()
-        .status(StatusCode::INTERNAL_SERVER_ERROR)
-        .body(INTERNAL_SERVER_ERROR.into())
-        .unwrap()
-}
-
 async fn simple_file_send(filename: &str) -> Result<Response<Body>> {
-    // Serve a file by asynchronously reading it entirely into memory.
-    // Uses tokio_fs to open file asynchronously, then tokio::io::AsyncReadExt
-    // to read into memory asynchronously.
+    // Serve a file by asynchronously reading it by chunks using tokio-util crate.
 
-    if let Ok(mut file) = File::open(filename).await {
-        let mut buf = Vec::new();
-        if let Ok(_) = file.read_to_end(&mut buf).await {
-            return Ok(Response::new(buf.into()));
-        }
-
-        return Ok(internal_server_error());
+    if let Ok(file) = File::open(filename).await {
+        let stream = FramedRead::new(file, BytesCodec::new());
+        let body = Body::wrap_stream(stream);
+        return Ok(Response::new(body));
     }
 
     Ok(not_found())

--- a/examples/send_file_index.html
+++ b/examples/send_file_index.html
@@ -3,9 +3,8 @@
     <title>Hyper responding example</title>
   </head>
   <body>
-    <h1>Hyper responding example</h1>
+    <h1>Hyper responding example, streamed in chunks</h1>
     <a href="index.html">index.html</a> Top Level<br>
-    <a href="big_file.html">big_file.html</a> This page, streamed in chunks<br>
     <a href="no_file.html">no_file.html</a> A 404 test, the requested file does not exist<br>
   </body>
 </html>

